### PR TITLE
Extend clangd with switch to header/source

### PR DESCRIPTION
**Pull request by petr-tik**
Address #2279 

Made an interactive function to switch to header/source.

Add a cmake-based repo for intergration tests of clangd. Currently Linux-only, since I don't have a Windows machine at home. 

I am not sure if async functions that fail inside deferred::nextc show up as errors in CI logs. Even if followed by deferred:sync! 

Follow up to #2279, which I closed to sort out the CI without spamming the upstream and reviewers.

### CODE REVIEW

**Comment by petr-tik** [comment]
I don't really understand why the windows ci run is failing. I don't think it's my clangd integration tests, but i don't think we even reach that point.

**Comment by petr-tik** [comment]
Can you please review again? @nbfalcon @yyoncho still don't understand why the windows CI is failing.

**Comment by yyoncho** [comment]
I am fine with the code. 

> still don't understand why the windows CI is failing.

My guess is that the current directory is not on the load-path thus it cannot find `lsp-integration-test.el`, it can be added explicitly to , I don't have an explanation why. You may try to add it explicitly to the load-path variable.

**Comment by petr-tik** [comment]
> My guess is that the current directory is not on the load-path thus it cannot find `lsp-integration-test.el`. I don't have an explanation why. You may try to add it explicitly to the load-path variable.

I think it's because of the alphabetical order. test/lsp-clangd-test.el is loaded before test/lsp-integration-test.el, so when i `(require 'lsp-integration-test)` in lsp-clangd-test.el, this file hadn't been loaded yet.

**Comment by petr-tik** [comment]
But then again, it somehow works in linux, so I am not sure

**Comment by petr-tik** [comment]
> I am fine with the code.

If that's still the case - please take a butcher's and feel free to squash and merge. 

Thanks for guiding me all along, appreciate it was longer than expected, but I now have more context to contribute other features.

**Comment by nbfalcon** [comment]
I think the windows CI being broken isn't your fault - the CI uses a command that was disabled for security resasons and wasn't updated yet.

File: clients/lsp-clangd.el
Comment (nbfalcon) [inline clients/lsp-clangd.el]: To be consistent with `projectile-find-other-file`, this should be named `lsp-clangd-find-other-file`.

File: clients/lsp-clangd.el
Comment (nbfalcon) [inline clients/lsp-clangd.el]: You cannot and shouldn't use `lsp-goto-location`, because that expects an `LspLocation` with line:col information, which is not what `clangd` returns here (it would also not make sense for it to do that).

**Review by nbfalcon — CHANGES_REQUESTED** [review]

File: clients/lsp-clangd.el
Comment (petr-tik) [reply clients/lsp-clangd.el]: Thank for addressing my TODOs. I will delete the comment.

Comment (nbfalcon) [reply clients/lsp-clangd.el]: Are there still any ToDos remaining?

File: clients/lsp-clangd.el
Comment (petr-tik) [reply clients/lsp-clangd.el]: Makes sense. Is consistency with projectile something that I should generally keep in mind? What about project.el that seems to be gaining feature parity?

**Comment by nbfalcon** [comment]
If @yyoncho is ok with that (don't do it yet), can you please squash your changes using interactive-rebase? There are currently too many commits.

**Comment by nbfalcon** [comment]
Things like `Wrote a free standing function` and the following two commits should be one. You are committing too often, IMHO.

**Comment by nbfalcon** [comment]
Your debugging commits like "try tmate" followed by "comment tmate out" should be gone as well.

File: clients/lsp-clangd.el
Comment (nbfalcon) [reply clients/lsp-clangd.el]: That is not something you need to keep in mind especially; we don't depend on `projectile`. This is just about having naming that is consistent with the rest of Emacs in general.

**Comment by nbfalcon** [comment]
I'd like to propose the following (tested) alternative implementation:
```
(defun lsp-clangd-find-other-file ()
  "Switch between the corresponding C/C++ source and header file.
Only works with `clangd'."
  (interactive)
  (let ((other (lsp-send-request (lsp-make-request
                                  "textDocument/switchSourceHeader"
                                  (lsp--text-document-identifier)))))
    (unless (s-present? other))
      (user-error "Could not find other file"))
    (find-file (lsp--uri-to-path other))))
```
My function has the following advantages:
- It doesn't break down if users have a custom `clangd` server id for tramp
- It `user-error`s instead of calling `lsp--info`, which gives a better hint that operation failed. 

It should be clear from both the naming and the docstring that this function cannot be used with anything but `clangd`, which is why I don't consider the clangd workspace check necessary. Servers that won't support this will error intuitively (IMHO):
- `jdtls`: "Unsupported request method: textDocument/switchSourceHeader"
- `pyright`: "Unhandled method textDocument/switchSourceHeader"

**Comment by nbfalcon** [comment]
@yyoncho do we prefer `lsp--info`, `lsp--error`, `message`, or `user-error`?

File: test/lsp-clangd-test.el
Comment (nbfalcon) [inline test/lsp-clangd-test.el:105]: Can you not do that? It should work on all platforms, so we should check on windows, too. I can have a look at the CI logs.

**Review by nbfalcon — CHANGES_REQUESTED** [review]

File: clients/lsp-clangd.el
Comment (yyoncho) [reply clients/lsp-clangd.el]: Generally, for the extension methods, we follow the name of the server method. In this case, something like lsp-clangd-switch-source-header should do.

**Comment by yyoncho** [comment]
> @yyoncho do we prefer `lsp--info`, `lsp--error`, `message`, or `user-error`?

For that particular use case, I would pick `user-error`. In general, I am less picky about non-core files.

**Review by yyoncho — APPROVED** [review]
Looks good to me! If @nbfalcon is fine we could merge it is as it is, to get the functionality in and then address the non-critical issues separately.

**Comment by petr-tik** [comment]
I can address the code review, function rename and rebase tidy up in the next couple of hours. Not sure about running clangd tests on windows CI though, as it’s been flaky for me and I don’t have a windows machine to test locally. 

I can add windows ci separately

**Comment by yyoncho** [comment]
> Not sure about running clangd tests on windows CI though, as it’s been flaky for me and I don’t have a windows machine to test locally.

Ideally, better to have these green but IMHO if you feel that you are spending too much time on it I am fine to not have these tests. After all, we don't that many resources to invest them in relatively low priority issues.

File: clients/lsp-clangd.el
Comment (nbfalcon) [reply clients/lsp-clangd.el]: @yyoncho but in this case I'd argue that there is benefit to be had from being consistent with the `projectile` naming scheme:
1. That name is good and intuitive (IMHO)
2. More familiarity: users of `projectile` could discover this function more easily trough `M-x`

**Comment by yyoncho** [comment]
> > Not sure about running clangd tests on windows CI though, as it’s been flaky for me and I don’t have a windows machine to test locally.
> 
> Ideally, better to have these green but IMHO if you feel that you are spending too much time on it I am fine to not have these tests. After all, we don't that many resources to invest them in relatively low priority issues.

.. especially given the fact that we dont have tests at all, so adding them only for Linux is still a huge gain.

**Comment by nbfalcon** [comment]
I might be able to test on windows.

**Comment by petr-tik** [comment]
> I'd like to propose the following (tested) alternative implementation:
> 
> ```emacs-lisp
> (defun lsp-clangd-find-other-file ()
>   "Switch between the corresponding C/C++ source and header file.
> Only works with `clangd'."
>   (interactive)
>   (let ((other (lsp-send-request (lsp-make-request
>                                   "textDocument/switchSourceHeader"
>                                   (lsp--text-document-identifier)))))
>     (unless (s-present? other))
>       (user-error "Could not find other file"))
>     (find-file (lsp--uri-to-path other))))
> ```

I was reviewing what you proposed vs what I have. I would quite like to keep the `(interactive "P") to open the file in another buffer. @yyoncho please be the tie breaker

**Comment by petr-tik** [comment]
> I might be able to test on windows.

Are you if we merge my PR without windows tests and you can add them later? To be honest, I don't really want to drag this relatively small change out any longer and think some tests is better than no tests.

**Comment by nbfalcon** [comment]
Here is a new version with prefix argument support:
```emacs-lisp
(defun lsp-clangd-find-other-file (new-window)
  "Switch between the corresponding C/C++ source and header file.
If NEW-WINDOW (interactively the prefix argument) is non-nil,
open the other file in a new window.

Only works with `clangd'."
  (interactive "P")
  (let ((other (lsp-send-request (lsp-make-request
                                  "textDocument/switchSourceHeader"
                                  (lsp--text-document-identifier)))))
    (unless (s-present? other)
      (user-error "Could not find other file"))
    (funcall (if new-window #'find-file-other-window #'find-file)
             (lsp--uri-to-path other))))
```

**Comment by nbfalcon** [comment]
I looked at the code, and I'd like you to do the following before it can be merged:
- [ ] use my revised method instead, for the reasons stated above
- [ ] create a new lsp-test-utils file, and move the common code there; in addition, refactor `lsp-with-pyls` to be generic, e.g. `lsp-with-workspace`, and define `lsp-clangd` and `lsp-with-pyls` in terms of the former
- [ ] (*optional*) make tests work on windows
- [ ] (optional) interactive-rebase
  - [ ] remove debugging commits
  - [ ] squash the free-standing implementation until implementation commits
  - [ ] squash the text fixture commits

**Comment by nbfalcon** [comment]
Note that my new `lsp-clangd-find-other-file` method `user-error`s; this means that you'll need to wrap calls to it for cases where the call should not switch the file in `should-error`, as otherwise the tests would fail.

**Comment by petr-tik** [comment]
>     * create a new lsp-test-utils file, and move the common code there; in addition, refactor `lsp-with-pyls` to be generic, e.g. `lsp-with-workspace`, and define `lsp-clangd` and `lsp-with-pyls` in terms of the former

Can you please create a separate issue for that and assign to me? I am not too proficient with elisp macros and would need a bit of time to work out how to make it generic over clangd and pyls considering they have different setup routines, which will drag this PR out, while extending outside the scope of the original ticket.

**Comment by petr-tik** [comment]
I have rebased to leave only 4 commits. Overall, I am pretty keen to have this merged and I will happily tidy up the integration testing code and get it to work for windows in another PR.

**Comment by petr-tik** [comment]
ok, for some reason my clangd tests aren't running anymore - pretty sure it's the tags in make unix-test

**Comment by petr-tik** [comment]
Sigh. I guess I will need to make it windows compatible after all.

File: test/lsp-clangd-test.el
Comment (nbfalcon) [inline test/lsp-clangd-test.el:147]:
```suggestion
      (should-error (lsp-clangd-find-other-file) :type 'user-error)
```

File: test/lsp-clangd-test.el
Comment (nbfalcon) [inline test/lsp-clangd-test.el:121]:
```suggestion
      (lsp-clangd-find-other-file)
```

File: test/lsp-clangd-test.el
Comment (nbfalcon) [inline test/lsp-clangd-test.el:129]:
```suggestion
      (lsp-clangd-find-other-file)
```

File: clients/lsp-clangd.el
Comment (nbfalcon) [inline clients/lsp-clangd.el:237]:
```suggestion
(defun lsp-clangd-find-other-file (&optional new-window)
```
Sorry, new-window should be &optional.

**Review by nbfalcon — CHANGES_REQUESTED** [review]
NEW-WINDOW should be optional.

**Comment by nbfalcon** [comment]
If you address my latest review, we can merge. This was a mistake on my part, sorry.

No windows support is good enough for now; I can look into that and do the macro refactoring in a later PR if you don't want to/can't do that.

Thank you for contributing!

**Comment by petr-tik** [comment]
> No windows support is good enough for now; I can look into that and do the macro refactoring in a later PR if you don't want to/can't do that.

Ok, I am back to fixing windows tests. 

> Thank you for contributing!

Thank you for helping me with this PR. It's been quite arduous for you - i appreciate that

**Comment by nbfalcon** [comment]
LGMT. I have tested this manually (`lsp-clangd-find-other-file` works) and the tests pass locally on my machine. This just needs to be rebased against master.

**Review by nbfalcon — APPROVED** [review]

**Comment by petr-tik** [comment]
@nbfalcon excuse the delay - i came back to this, rebased on master (looks like you added macos tests - nice!) and it looks fine bar the windows test fails, which I thought you were ok with? Please let me know what I should do

**Comment by nbfalcon** [comment]
Great! Though note that it wasn't me who has added the `macos` tests, I think it was @jcs090218. I can't merge this with failing tests, so @yyoncho will need to do that.

**Comment by yyoncho** [comment]
> Great! Though note that it wasn't me who has added the `macos` tests, I think it was @jcs090218. I can't merge this with failing tests, so @yyoncho will need to do that.

Are we going to merge with broken tests? If you are planning to fix them, can we comment on them?

**Comment by kiennq** [comment]
I'm not a fan of broken tests since it will make us not sure if new changes will broken the package further.
Can we at least has the `system-type` check so the tests for these new functionalities will run only when is not in `windows-nt` or `darwin`?

**Comment by nbfalcon** [comment]
I agree, broken CI's don't seem like a good idea. Perhaps something like `vagrant` might help with local cross-platform testing? Now the tests can be fixed by wrapping them in `(unless (memq system-type '(windows-nt darwin)))`.

**Comment by petr-tik** [comment]
skipping clangd integration tests unless on linux atm. macos seems to fail to install dependencies.

**Review by kiennq — APPROVED** [review]

**Review by yyoncho — APPROVED** [review]

**Comment by yyoncho** [comment]
Thank you!

**Pull request merged**
